src/compiler/parser.ts

```diff
@@ -5231,7 +5231,7 @@ module ts {
                 }
             }
             else {
-                if (!(findSourceFile(filename + ".ts", isDefaultLib, refFile, refPos, refEnd) || findSourceFile(filename + ".d.ts", isDefaultLib, refFile, refPos, refEnd))) {
+                if (!((options.allowNonTsExtensions && findSourceFile(filename, isDefaultLib, refFile, refPos, refEnd)) || findSourceFile(filename + ".ts", isDefaultLib, refFile, refPos, refEnd) || findSourceFile(filename + ".d.ts", isDefaultLib, refFile, refPos, refEnd))) {

```

Comment:
Can you break this up into individual checks.  Trying to read through this complex conditional is somewhat unpleasant.

(or, just put the individual cases on individual lines.  that alone might be enough to make this readable).